src/main/java/org/zalando/nakadi/repository/MultiTimelineEventConsumer.java

```diff
@@ -0,0 +1,265 @@
+            }
+            newAssignment.get(topicRepository).add(Optional.ofNullable(cursorReplacement.get()).orElse(cursor));
+        }
+        final Set<TopicRepository> removedTopicRepositories = eventConsumers.keySet().stream()
+                .filter(tr -> !newAssignment.containsKey(tr))
+                .collect(Collectors.toSet());
+        // Stop and remove event consumers that are not needed anymore
+        for (final TopicRepository toRemove : removedTopicRepositories) {
+            stopAndRemoveConsumer(toRemove);
+        }
+        // Stop and remove event consumers with changed configuration
+        for (final Map.Entry<TopicRepository, List<NakadiCursor>> entry : newAssignment.entrySet()) {
+            final EventConsumer existingEventConsumer = eventConsumers.get(entry.getKey());
+            if (null != existingEventConsumer) {
+                final Set<TopicPartition> newTopicPartitions = entry.getValue().stream()
+                        .map(NakadiCursor::getTopicPartition)
+                        .collect(Collectors.toSet());
+                if (!existingEventConsumer.getAssignment().equals(newTopicPartitions)) {
+                    stopAndRemoveConsumer(entry.getKey());
+                }
+            }
+        }
+        // Start new consumers with changed configuration.
+        for (final Map.Entry<TopicRepository, List<NakadiCursor>> entry : newAssignment.entrySet()) {
+            if (!eventConsumers.containsKey(entry.getKey())) {
+                final TopicRepository repo = entry.getKey();
+                final EventConsumer consumer = repo.createEventConsumer(clientId, entry.getValue());
+                eventConsumers.put(repo, consumer);
+            }
+        }
+    }
+
+    private void stopAndRemoveConsumer(final TopicRepository toRemove) {
+        final EventConsumer realConsumer = eventConsumers.remove(toRemove);
+        try {
+            realConsumer.close();
+        } catch (IOException ex) {
+            LOG.error("Failed to stop one of consumers, but will not care about that, " +
+                    "because it is already doesn't matter", ex);
+        }
+    }
+
+
+    private void timelinesChanged() throws NakadiException, InvalidCursorException {
+        final Set<String> eventTypes = latestOffsets.values().stream()
+                .map(NakadiCursor::getEventType)
+                .collect(Collectors.toSet());
+        // Remove obsolete data
+        final List<String> toRemove = consumedEventTypes.keySet().stream()
+                .filter(et -> !eventTypes.contains(et))
+                .collect(Collectors.toList());
+        for (final String item : toRemove) {
+            consumedEventTypes.remove(item);
+        }
+        // Add refreshed data.
+        for (final String eventType : eventTypes) {
+            consumedEventTypes.put(eventType, timelineService.getActiveTimelinesOrdered(eventType));
+        }
```

Comment:
Ah, ok I see. Because of subscription rebalance the set of streamed event-types can be changed, right?